quarkus/addons/token-exchange/runtime/src/main/java/org/kie/kogito/addons/quarkus/token/exchange/cache/TokenEvictionHandler.java

```diff
@@ -0,0 +1,120 @@
+     */
+    public RemovalListener<String, CachedTokens> createRemovalListener() {
+        return (key, value, cause) -> {
+            if (value == null)
+                return;
+
+            LOGGER.info("Token cache eviction for cache key '{}' - Cause: {}", key, cause);
+            onTokenExpired(key, value, cause);
+        };
+    }
+
+    /**
+     * Callback method called when tokens are evicted from the cache.
+     * 
+     * @param cacheKey The cache key of the evicted tokens
+     * @param tokens The evicted token data
+     * @param cause The reason for eviction (Caffeine's RemovalCause)
+     */
+    private void onTokenExpired(String cacheKey, CachedTokens tokens, RemovalCause cause) {
+        LOGGER.warn("OAuth2 tokens for cache key '{}' have expired/been evicted: {}", cacheKey, cause);
+
+        // Handle proactive token refresh when cache entry expires (which happens before actual token expiration)
+        if (cause == RemovalCause.EXPIRED) {
+            if (tokens.refreshToken() == null) {
+                LOGGER.warn("OAuth2 tokens for cache key '{}' has no refresh token, the access token cannot be refreshed.", cacheKey);
+                return;
+            }
+
+            if (!tokens.isExpiredNow()) {
+                LOGGER.info("Triggering proactive token refresh for cache key '{}' (token still valid but refresh window reached)", cacheKey);
+                refreshWithCachedToken(cacheKey, tokens.refreshToken());
+            }
+        } else if (cause == RemovalCause.EXPLICIT) {
+            LOGGER.info("Deleting OAuth2 tokens with cache key '{}' from persistence system", cacheKey);
+            tokenCRUD.deleteToken(cacheKey);
+        }
+    }
+
+    /**
+     * Refreshes tokens using a cached refresh token and updates the cache.
+     * 
+     * @param cacheKey The cache key for the tokens being refreshed
+     * @param refreshToken The refresh token to use for getting new tokens
+     */
+    private void refreshWithCachedToken(String cacheKey, String refreshToken) {
+        CompletableFuture.runAsync(() -> {
+            try {
+                LOGGER.info("{} - cache key '{}'", LOG_PREFIX_TOKEN_REFRESH, cacheKey);
+
+                String authName = CacheUtils.extractAuthNameFromCacheKey(cacheKey);
+                OidcClient client = OidcClientUtils.getExchangeTokenClient(authName);
+
+                LOGGER.debug("Refreshing token for cache key '{}' using cached refresh token", cacheKey);
+
+                Tokens refreshedTokens = client.getTokens(Collections.singletonMap("refresh_token", refreshToken))
+                        .await().indefinitely();
+
+                tokenCRUD.storeToken(cacheKey, refreshedTokens);
+
+                LOGGER.info("{} - cache key '{}'", LOG_PREFIX_REFRESH_COMPLETED, cacheKey);
+            } catch (Exception e) {
+                LOGGER.error("{} - cache key '{}': {}", LOG_PREFIX_FAILED_TO_REFRESH_TOKEN, cacheKey, e.getMessage());
```

Comment:
you could pass the complete stacktrace, not only the message
```suggestion
                LOGGER.error("{} - cache key '{}': {}", LOG_PREFIX_FAILED_TO_REFRESH_TOKEN, cacheKey, e);
```